src/bp/core/services/bot-service.ts

```diff
@@ -451,4 +465,75 @@ export class BotService {
     BotService._mountedBots.forEach((isMounted, bot) => isMounted && bots.push(bot))
     return bots
   }
+
+  public async listRevisions(botId: string): Promise<string[]> {
+    const globalGhost = this.ghostService.global()
+
+    let stageID = ''
+    if (await this.workspaceService.hasPipeline()) {
+      const botConfig = await this.configProvider.getBotConfig(botId)
+      stageID = botConfig.pipeline_status.current_stage.id
+    }
+
+    const revisions = await globalGhost.directoryListing(REVISIONS_DIR)
+
+    return revisions.filter(rev => rev.includes(botId) && rev.includes(stageID)).sort((revA, revB) => {
+      const dateA = revA.split('::')[1].replace('.tgz', '')
+      const dateB = revB.split('::')[1].replace('.tgz', '')
+      const diff = moment(dateA, 'YY-MM-DD-ha').diff(moment(dateB, 'YY-MM-DD-ha'))
+      return diff * -1 // descending
+    })
+  }
+
+  public async createRevision(botId: string): Promise<void> {
+    const botConfig = await this.configProvider.getBotConfig(botId)
+    let revName = `${botId}::${moment().format('YY-MM-DD-ha')}`
+
+    if (await this.workspaceService.hasPipeline()) {
+      revName = `${revName}::${botConfig.pipeline_status.current_stage.id}`
+    }
+
+    const botGhost = this.ghostService.forBot(botId)
+    const globalGhost = this.ghostService.global()
+    await globalGhost.ensureDirs('/', ['revisions'])
+    await globalGhost.upsertFile(REVISIONS_DIR, revName + '.tgz', await botGhost.exportToArchiveBuffer())
+    return this._cleanupRevisions(botId)
+  }
+
+  public async rollback(botId: string, revision: string): Promise<void> {
+    const botConfig = await this.configProvider.getBotConfig(botId)
```

Comment:
Same here, this bot config can be checked in the if(hasPipeline) below, to avoid an unneeded call if any of the below exceptions are thrown